### PR TITLE
Configure Debian repositories on the bootstrap

### LINKED ISSUE #1726: Configure Debian repository on bootstrap node
**Component**: Bootstrap

**Why this is needed**: To deploy a bootstrap on an Ubuntu node, we need to configure our Debian repository to provide the required packages.

**What should be done**: 
Modify the bootstrap.sh.in script in the buildchain to configure the Debian repository in local for the Ubuntu bootstrap node according to the os.

**Implementation proposal** (strongly recommended):
implement a solution to be sur that the script do the os selection correctly .
Create a function for Debian repository configuration on an Ubuntu bootstrap node according to the os.

**Test plan**: Launch the bootstrap deployment on an Ubuntu node then verify our Debian repository is well configured.

### LINKED ISSUE #1727: Install Salt-minion on the Ubuntu bootstrap node
**Component**: Salt, bootstrap

**Why this is needed**: We need to permit to the bootstrap to install the salt-minion 

**What should be done**: Modify the bootstrap script to allow the minion installation on the Ubuntu bootstrap node according to the OS

**Implementation proposal** (strongly recommended):

**Test plan**:
Launch the bootstrap scrip on an Ubuntu machine then verify that the salt-minion is well installed

### PR DESCRIPTION
**Pull request by ycointe**
**Component**: Deployment, Bootstrap

**Context**: MetalK8s on Ubuntu

**Summary**: We want to have the possibility to deploy the bootstrap node on a Ubuntu machine.
We need to deploy the Debian repository localy to permit the installation offline.

**Acceptance criteria**: 
- Launch the bootstrap script on a Ubuntu machine 
- Verify that you have all the required `metalk8s-*.list` in `/etc/apt/sources.list/`

---

Closes: #1726 
Closes: #1727

### CODE REVIEW

File: scripts/bootstrap.sh.in [hunk 1]
```diff
@@ -53,11 +65,14 @@ trap cleanup EXIT
 BASE_DIR=$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")
 
 RPM=${RPM:-$(command -v rpm)}
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: Need to be replaced by:
```RPM=${RPM:-$(command -v rpm || true)}```
Otherwise it'll fail.
Same for dpkg, apt & yum.

File: scripts/bootstrap.sh.in [hunk 2]
```diff
@@ -147,6 +187,19 @@ disable_salt_minion_service() {
 
 stop_salt_minion_service() {
     ${SYSTEMCTL} stop salt-minion.service 2>/dev/null || true
+
+configure_repositories() {
+
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: Extra line

File: scripts/bootstrap.sh.in [hunk 3]
```diff
@@ -67,6 +82,15 @@ declare -A GPGCHECK_REPOSITORIES=(
     [metalk8s-scality]=0
 )
 
+declare -A GPGCHECK_APT_REPOSITORIES=(
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in:91]: The values of your hash don't seem to be used anywhere.

File: scripts/bootstrap.sh.in [hunk 4]
```diff
@@ -75,7 +99,7 @@ run_quiet() {
     local start
     start=$(date +%s)
     set +e
-    "$@" 2>&1 | tee -ia "${LOGFILE}" > "${TMPFILES}/out"
+    "$@" > >(tee -ia "${LOGFILE}" > "${TMPFILES}/out") 2>&1
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in:108]: Commit message is wrong, it's not related to Ubuntu.
It's because you set global variables in one of the command executed here.
Should rather be something like:
```
Replace pipe by a process substitution to allow global variable assignments in the command ran.
```

File: scripts/bootstrap.sh.in [hunk 5]
```diff
@@ -182,24 +248,51 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+            local -a yum_opts=(
+                '--assumeyes'
+                '--disablerepo=*'
+                '--enablerepo=metalk8s-*'
+            )
+
+            "$YUM" install "${yum_opts[@]}" salt-minion
+            ;;
+        ubuntu)
+            "$APT" install salt-minion
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: It's missing option to apt, otherwise it'll be launched in interactive mode.

File: scripts/bootstrap.sh.in [hunk 6]
```diff
@@ -182,24 +248,51 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+            ;;
+        *)
+            die "Impossible to install Salt-minion";;
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: Error message is wrong, it's not impossible to install salt-minion, but the OS is not supported by our script.

File: scripts/bootstrap.sh.in [hunk 7]
```diff
@@ -182,24 +248,51 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+                '--enablerepo=metalk8s-*'
+             )
+
+            "$YUM" install "${yum_opts[@]}" genisoimage
+            ;;
+        ubuntu)
+            "$APT" install genisoimage
+            ;;
+        *)
+            die "Impossible to install genisoimage";;
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: Same as above, wrong error message.
Maybe... instead of having this case everywhere, you could check once and for all that we support the current OS at the beginning of the script.

File: scripts/bootstrap.sh.in [hunk 8]
```diff
@@ -182,24 +248,51 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
-        '--assumeyes'
-        '--disablerepo=*'
-        '--enablerepo=metalk8s-*'
-    )
+configure_apt_local_repository() {
+    local -r repo_name=$1
+    local -r repo_path="$BASE_DIR/packages/debian/$repo_name"
+
+    echo "deb [trusted=yes] file://$repo_path bionic $repo_name" \
+        >   /etc/apt/sources.list.d/"$repo_name".list
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: details: some extra spaces after `>`

File: scripts/bootstrap.sh.in [hunk 9]
```diff
@@ -182,24 +248,51 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: extra line

File: scripts/bootstrap.sh.in [hunk 10]
```diff
@@ -44,6 +46,16 @@ EOF
 
 exec > >(tee -ia "${LOGFILE}") 2>&1
 
+determine_os() {
+    # This way permit to determine All the recent OS like Centos, Ubuntu
+    # and Debian
+     if [ -f /etc/os-release ]; then
+         . /etc/os-release;
+         OS=$ID;
+         VER=$VERSION_ID;
```
Comment (slaperche-scality) [inline scripts/bootstrap.sh.in]: VER is a global variable?
If yes:
- use a longer name (`OS_VERSION` maybe)
- declare it like `OS`, at the top of the script.

File: scripts/bootstrap.sh.in [hunk 11]
```diff
@@ -147,6 +187,19 @@ disable_salt_minion_service() {
 
 stop_salt_minion_service() {
     ${SYSTEMCTL} stop salt-minion.service 2>/dev/null || true
+
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: it's missing a closing curly brace.

File: scripts/bootstrap.sh.in [hunk 12]
```diff
@@ -44,6 +46,16 @@ EOF
 
 exec > >(tee -ia "${LOGFILE}") 2>&1
 
+determine_os() {
+    # This way permit to determine All the recent OS like Centos, Ubuntu
```
Comment (slaperche-scality) [inline scripts/bootstrap.sh.in]: I would reword as `we rely on /etc/os-release to discover the OS because its present on all recent Linux distributions`

File: scripts/bootstrap.sh.in (hunk 10 above)
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: This variable does not seem to be used anywhere.
details: useless semicolons at the end of your 3 lines above.

File: scripts/bootstrap.sh.in [hunk 13]
```diff
@@ -65,8 +65,10 @@ trap cleanup EXIT
 BASE_DIR=$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")
 
 RPM=${RPM:-$(command -v rpm)}
-SYSTEMCTL=${SYSTEMCTL:-$(command -v systemctl)}
+DPKG=${DPKG:-$(command -v dpkg)}
```
Comment (slaperche-scality) [inline scripts/bootstrap.sh.in]: in commit message: `Integrate tests for Ubuntu ` => `add pre-checks for Ubuntu`

File: scripts/bootstrap.sh.in [hunk 14]
```diff
@@ -143,14 +145,30 @@ die() {
     return 1
 }
 
+os_dependent_checks() {
```
Comment (slaperche-scality) [inline scripts/bootstrap.sh.in]: `check_package_managers`?

File: scripts/bootstrap.sh.in [hunk 15]
```diff
@@ -187,6 +187,19 @@ disable_salt_minion_service() {
 
 stop_salt_minion_service() {
     ${SYSTEMCTL} stop salt-minion.service 2>/dev/null || true
+
```
Comment (slaperche-scality) [inline scripts/bootstrap.sh.in]: In commit message: `redHat` => `RedHat`

File: scripts/bootstrap.sh.in [hunk 16]
```diff
@@ -44,6 +46,16 @@ EOF
 
 exec > >(tee -ia "${LOGFILE}") 2>&1
 
+determine_os() {
+    # This way permit to determine All the recent OS like Centos, Ubuntu
+    # and Debian
+     if [ -f /etc/os-release ]; then
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: What if this file does not exist ? No fallback ? No error ?
To me, you should at least return a non-zero exit code with an explicit error message to make the script fail.

File: scripts/bootstrap.sh.in [hunk 17]
```diff
@@ -187,6 +187,19 @@ disable_salt_minion_service() {
...
+
+    case "$OS" in
+        centos)
+            configure_yum_repositories
+            ;;
+        ubuntu)
+            configure_apt_repositories
+            ;;
+        *)
+            die "OS found not supported: $OS";;
```
Comment (slaperche-scality) [inline scripts/bootstrap.sh.in]: `OS $OS not supported`

File: scripts/bootstrap.sh.in (hunk 4 above)
Comment (slaperche-scality) [reply scripts/bootstrap.sh.in:108]: Still in commit message: `it occurs` => `it creates`

Could be worth to have a comment here on addition to the commit message actually.

File: scripts/bootstrap.sh.in (hunk 1 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: Indeed, to avoid the fails for my tests, I comment the `set -e` thank you for the reminder !

File: scripts/bootstrap.sh.in (hunk 5 above)
Comment (jbertran) [reply scripts/bootstrap.sh.in]: The actual command should be something like `DEBIAN_FRONTEND=noninteractive apt install -y`

File: scripts/bootstrap.sh.in (hunk 10 above)
Comment (jbertran) [reply scripts/bootstrap.sh.in]: IIRC this has no effect yet, it's only exported in case we ever need to switch on versions, so we can safely remove it until we require it.

Comment (ycointe) [reply scripts/bootstrap.sh.in]: Indeed, I put it in case of we support several version of one OS, it permit to sort by OS version, but it's not accurate for now... Should I deleted it ?

File: scripts/bootstrap.sh.in [hunk 18]
```diff
@@ -182,24 +248,51 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+                '--disablerepo=*'
+                '--enablerepo=metalk8s-*'
+             )
+
+            "$YUM" install "${yum_opts[@]}" genisoimage
+            ;;
+        ubuntu)
+            "$APT" install genisoimage
+            ;;
+        *)
```
Comment (TeddyAndrieux) [inline scripts/bootstrap.sh.in:294]: Maybe having a generic function to "install" a package make sense even if we only have this 2 package for the moment

File: scripts/bootstrap.sh.in (hunk 16 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: from what I read about it, all new system integrate it, it's not present on all other oldest system, but always here on the new ones.

File: scripts/bootstrap.sh.in (hunk 10 above)
Comment (alexandre-allard) [reply scripts/bootstrap.sh.in]: If it's not useful in any way, it should be deleted, yes.
We will add it back the day we'll need it... if this day comes.

File: scripts/bootstrap.sh.in (hunk 16 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: I add a `die "Impossible to determine the OS"`

File: scripts/bootstrap.sh.in (hunk 10 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: I deleted VER variable because of it's not used

File: scripts/bootstrap.sh.in (hunk 12 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: Better than my message indeed ! :)

File: scripts/bootstrap.sh.in (hunk 14 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: looks better ! :+1:

File: scripts/bootstrap.sh.in (hunk 10 above)
Comment (ycointe) [reply scripts/bootstrap.sh.in]: I deleted it.

File: scripts/bootstrap.sh.in (hunk 16 above)
Comment (alexandre-allard) [reply scripts/bootstrap.sh.in]: If we want to fully use run_quiet(), you should rather do something like:
```
echo "Impossible to determine the OS" 1>&2
return 1
```

File: scripts/bootstrap.sh.in [hunk 19]
```diff
@@ -182,24 +254,31 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+        centos)
+            local -a yum_opts=(
+                '--assumeyes'
+                '--disablerepo=*'
+                '--enablerepo=metalk8s-*'
+            )
+            "$YUM" install "${yum_opts[@]}" "${packages[@]}"
+            ;;
+        ubuntu)
+            DEBIAN_FRONTEND=noninteractive "$APT" -y install "${packages[@]}"
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: Not sure if we want to do it also on Ubuntu but, on CentOS we disable all repositories that are not `metalk8s-*` when installing the packages.
I know there is no clean way to do it, but we can do something like:
```
local -a repo_list=(/etc/apt/sources.list)

mapfile -O 1 -t repo_list < <(find /etc/apt/sources.list.d/ ! -name 'metalk8s-*' -a -name '*.list')

for source_list in "${repo_list[@]}"; do
    mv "$source_list"{,.disabled}
done

## Install your packages here

for source_list in "${repo_list[@]}"; do
    mv "$source_list"{.disabled,}
done
```

Comment (ycointe) [reply scripts/bootstrap.sh.in]: Done, thank you ! I done some tests on Kubernetes and it was well downloaded from the new internal repo list, but it's more safe like this !

Comment (alexandre-allard) [reply scripts/bootstrap.sh.in]: Sorry, one last thing I should have told you.
You should prevent the script from failing here, otherwise you could end up with system repositories in a bad state (it should not, but hey better be safe than sorry).

File: scripts/bootstrap.sh.in [hunk 20]
```diff
@@ -182,24 +254,47 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+            "$YUM" install "${yum_opts[@]}" "${packages[@]}"
+            ;;
+        ubuntu)
+            mapfile -O 1 -t repo_list < <(find \
+                /etc/apt/sources.list.d/ \
+                ! -name 'metalk8s-*' -a -name '*.list')
+
+            for source_list in "${repo_list[@]}"; do
+                mv "$source_list"{,.disabled}
+            done
```
Comment (TeddyAndrieux) [inline scripts/bootstrap.sh.in:283]: Should you not `apt update` just after doing that kind of change ?
Also here you not disable all repos in `/etc/apt/sources.list` and we have a difference here between CentOs and Debian as for Debian our repos are always enabled. But I think we can keep like this (at least for the moment :smile: )

File: scripts/bootstrap.sh.in [hunk 21]
```diff
@@ -182,24 +254,48 @@ ${gpg_keys:+gpgkey=${gpg_keys%?}}
...
+            "$APT" update
+            DEBIAN_FRONTEND=noninteractive "$APT" -y install "${packages[@]}" \
+                || exit_code=$?
+
+            for source_list in "${repo_list[@]}"; do
+                mv "$source_list"{.disabled,}
+            done
+            "$APT" update
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: You should not update after restoring the repositories as you don't know if system repositories are well configured.
If it's not the case, the bootstrap will fail here because `apt update` will return a non-zero exit code.
Plus, even if repositories are well configured or if we catch the exit code, it'll take some time for something we don't care/need.

File: scripts/bootstrap.sh.in [hunk 22]
```diff
@@ -155,11 +214,24 @@ configure_yum_repositories() {
...
-            "${GPGCHECK_REPOSITORIES[$repository]}"
+            "${GPGCHECK_YUM_REPOSITORIES[$repository]}"
+    done
+}
+
+configure_apt_local_repositories() {
+    for repository in "${!GPGCHECK_APT_REPOSITORIES[@]}"; do
+        configure_apt_local_repository "$repository"
     done
+    "$APT" update
```
Comment (alexandre-allard) [inline scripts/bootstrap.sh.in]: You should not.

**Comment by alexandre-allard** [comment]
@ycointe waiting for you to apply some last changes and I think we're good.

**Comment by ycointe** [comment]
Done

**Review by alexandre-allard — APPROVED** [review]
